### PR TITLE
Support null_policy::EXCLUDE for COLLECT rolling aggregation

### LINKED ISSUE #7258: [FEA] COLLECT window aggregation should support null_policy::EXCLUDE
#7189 implements `COLLECT` aggregations to be done from window functions. The semantics of how null input rows are handled are consistent with CUDF semantics.
E.g. 
```c++
auto input_col = fixed_width_column_wrapper<int32_t>{70, ∅, 72, 73, 74};
auto output_col = cudf::rolling_window(input_col, 2, 1, 1, collect_aggr);
            // == [ [70,∅], [70,∅,72], [∅,72,73], [72,73,74], [73,74] ]
```
Note that the null element (`∅`) is replicated in the first 3 rows of the output.

SparkSQL (and Hive, and other big data SQL systems) have different semantics, in that all null elements are purged. The output for the same operation should yield the following:
```c++
auto sparkish_output_col = cudf::rolling_window(input_col, 2, 1, 1, collect_aggr);
            // == [ [70], [70,72], [72,73], [72,73,74], [73,74] ]
```

CUDF should allow the `COLLECT` aggregation to be constructed with an optional `null_policy` argument (with default `INCLUDE`). The `COLLECT` window function should check the policy, and filter out null list-elements _a posteriori_.

### PR DESCRIPTION
**Pull request by mythrocks**
Closes #7258.

#7189 implements `COLLECT` aggregations to be done from window functions. The semantics of how null input rows are handled are consistent with CUDF semantics.
E.g. 
```c++
auto input_col = fixed_width_column_wrapper<int32_t>{70, ∅, 72, 73, 74};
auto output_col = cudf::rolling_window(input_col, 2, 1, 1, collect_aggr);
            // == [ [70,∅], [70,∅,72], [∅,72,73], [72,73,74], [73,74] ]
```
Note that the null element (`∅`) is replicated in the first 3 rows of the output.

SparkSQL (and Hive, and other big data SQL systems) have different semantics, in that all null elements are purged. The output for the same operation should yield the following:
```c++
auto sparkish_output_col = cudf::rolling_window(input_col, 2, 1, 1, collect_aggr);
            // == [ [70], [70,72], [72,73], [72,73,74], [73,74] ]
```

CUDF should allow the `COLLECT` aggregation to be constructed with an optional `null_policy` argument (with default `INCLUDE`). The `COLLECT` window function should check the policy, and filter out null list-elements _a posteriori_.

### CODE REVIEW

**Comment by mythrocks** [comment]
The build failure might be because `branch-0.19` is quite new:
```
22:41:05 PackagesNotFoundError: The following packages are not available from current channels:
22:41:05 
22:41:05   - ucx-py=0.19
22:41:05   - rapids-build-env=0.19
22:41:05   - rapids-notebook-env=0.19
22:41:05 
22:41:05 Current channels:...
```

File: cpp/src/rolling/rolling_detail.cuh [hunk 1]
```diff
@@ -1062,6 +1062,72 @@ struct rolling_window_launcher {
     return gather_map;
   }
 
+  /**
+   * @brief Purge entries for null inputs from gather_map, and adjust offsets.
+   */
+  void purge_null_entries(column_view const& input,
+                          std::unique_ptr<column>& gather_map,
+                          std::unique_ptr<column>& offsets,
```
Comment (jrhemstad) [inline cpp/src/rolling/rolling_detail.cuh:1070]: Passing a `unique_ptr` by non-const reference declares that you intend to reseat that pointer. That doesn't appear to be the case here.

https://www.modernescpp.com/index.php/c-core-guidelines-passing-smart-pointer

File: cpp/src/rolling/rolling_detail.cuh [hunk 2]
```diff
@@ -1062,6 +1062,72 @@ struct rolling_window_launcher {
...
+      return d_input.is_null_nocheck(i);
+    };
+    auto input_row_not_null = [d_input = *input_device_view] __device__(auto i) {
+      return d_input.is_valid_nocheck(i);
+    };
+
+    auto num_child_nulls = thrust::count_if(rmm::exec_policy(stream),
+                                            gather_map->view().template begin<size_type>(),
+                                            gather_map->view().template end<size_type>(),
+                                            input_row_is_null);
```
Comment (jrhemstad) [inline cpp/src/rolling/rolling_detail.cuh:1086]: This looks like a much less efficient version of `column::null_count`. Can't this all be replaced by `input.null_count()`?

File: cpp/src/rolling/rolling_detail.cuh [hunk 3]
```diff
@@ -1062,6 +1062,72 @@ struct rolling_window_launcher {
...
+                      thrust::make_counting_iterator<size_type>(input.size()),
+                      new_sizes->mutable_view().template begin<size_type>(),
+                      [d_gather_map  = gather_map->view().template begin<size_type>(),
+                       d_old_offsets = offsets->view().template begin<size_type>(),
+                       input_row_not_null] __device__(auto i) {
+                        return thrust::count_if(thrust::seq,
+                                                d_gather_map + d_old_offsets[i],
+                                                d_gather_map + d_old_offsets[i + 1],
+                                                input_row_not_null);
+                      });
```
Comment (jrhemstad) [inline cpp/src/rolling/rolling_detail.cuh:1119]: This looks like it could likewise be replaced by `segmented_count_unset_bits` (with some modifications) https://docs.rapids.ai/api/libcudf/nightly/group__column__nullmask.html#ga81838307b321a8a911054944a74f4bfb

**Review by jrhemstad — CHANGES_REQUESTED** [review]

File: cpp/src/rolling/rolling_detail.cuh (hunk 1 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1070]: The `gather_map` and `offsets` are reseated at lines 1128-1129:
```c++
    ...
    gather_map = std::move(new_gather_map);
    offsets    = std::move(new_offsets);
```

File: cpp/src/rolling/rolling_detail.cuh (hunk 2 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1086]: > much less efficient version of `column::null_count`.

Do we mean `column_view::null_count(begin, end)`? That's brilliant. Let me try switching to that.

File: cpp/src/rolling/rolling_detail.cuh (hunk 3 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1119]: Thanks for the tip. I'll try rephrase this to use `segmented_count_set_bits()`.

File: cpp/src/rolling/rolling_detail.cuh (hunk 2 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1086]: > Can't this all be replaced by `input.null_count()`?

For the record, `input.null_count()` is already checked in the calling function. The check at Line#1088 is to see if the `gather_map` chooses any null elements from `input`. It doesn't look like `column::null_count()` or `column_view::null_count(begin,end)` makes that available.

There might be value in using `segmented_count_set_bits()`. I'm checking now.

Comment (jrhemstad) [reply cpp/src/rolling/rolling_detail.cuh:1086]: Ah, I see, you're doing a fused gather + null_count. I see now.

File: cpp/src/rolling/rolling_detail.cuh (hunk 3 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1119]: `segmented_count_set_bits()` takes a `std::vector` of segment offset (pairs). This implies that the vector would have nearly twice as many elements as there are list rows in the column. 
Pardon my naiveté, but is it permitted to copy the offsets column over to host memory (and doubled) to use `segmented_count_unset_bits()`?

Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1119]: Actually, no, nix that.
For the same reason as [described above](https://github.com/rapidsai/cudf/pull/7264#discussion_r568112965), it doesn't look like the offset recomputation can be expressed as a call to `segmented_count_set_bits()`, without it being convoluted. :/
In short, we're not counting valid rows in the `gather_map`; we're counting rows in the `gather_map` that point to valid rows in the `input` column.

**Comment by mythrocks** [comment]
I just checked that there isn't a pressing need to similarly exclude `NaN` values from a collected list, at least from SparkSQL. Spark retains `NaN` values.

**Comment by mythrocks** [comment]
rerun tests

File: cpp/src/rolling/rolling_detail.cuh (hunk 1 above)
Comment (jrhemstad) [reply cpp/src/rolling/rolling_detail.cuh:1070]: Ah I see. Any reason the old gather map/offsets can't be an input and the new ones be an output? That would be more canonical with libcudf style.

File: cpp/src/groupby/sort/groupby.cu [hunk 4]
```diff
@@ -403,6 +403,11 @@ void store_result_functor::operator()<aggregation::NTH_ELEMENT>(aggregation cons
 template <>
 void store_result_functor::operator()<aggregation::COLLECT>(aggregation const& agg)
 {
+  auto null_handling =
+    static_cast<cudf::detail::collect_list_aggregation const&>(agg)._null_handling;
+  CUDF_EXPECTS(null_handling == null_policy::INCLUDE,
```
Comment (rgsl888prabhu) [inline cpp/src/groupby/sort/groupby.cu:408]: If we are supporting `null_policy::EXCLUDE`, why is this expectation?

File: cpp/tests/collect_list/collect_list_test.cpp [hunk 5]
```diff
@@ -466,6 +636,87 @@ TYPED_TEST(TypedCollectListTest, BasicGroupedTimeRangeRollingWindow)
...
+                                      cudf::order::ASCENDING,
+                                      input_column,
+                                      preceding,
+                                      following,
+                                      min_periods,
+                                      make_collect_aggregation());
+
+  auto null_at = [](auto null_idx) {
+    return make_counting_transform_iterator(0, [null_idx](auto i) { return i != null_idx; });
+  };
```
Comment (vuule) [inline cpp/tests/collect_list/collect_list_test.cpp:680]: This is neat. IMO it would be nice to have it as a utility function (`iterator_with_null_at` or something like that).
A version that takes a vector of values (not that fast, but it's meant for small input AFAICT) could also be added.

**Review by rgsl888prabhu — COMMENTED** [review]
Jake has almost covered all essentials, have only minor question.

File: cpp/src/rolling/rolling_detail.cuh [hunk 6]
```diff
@@ -1105,6 +1171,13 @@ struct rolling_window_launcher {
     auto gather_map = create_collect_gather_map(
       offsets->view(), per_row_mapping->view(), preceding_begin, stream, mr);
 
+    // If gather_map collects null elements, and null_policy == EXCLUDE,
+    // those elements must be filtered out, and offsets recomputed.
+    auto null_handling = static_cast<collect_list_aggregation*>(agg.get())->_null_handling;
+    if (null_handling == null_policy::EXCLUDE && input.has_nulls()) {
+      purge_null_entries(input, gather_map, offsets, stream, mr);
+    }
```
Comment (vuule) [inline cpp/src/rolling/rolling_detail.cuh:1179]: Is the gather map here guaranteed to include all elements of the input table?
If so, we don't need the complex null check to short-circuit purge_null_entries - any null in input column is guaranteed to affect the output.

**Review by vuule — CHANGES_REQUESTED** [review]
Just a few suggestions, already looks great!

File: cpp/tests/collect_list/collect_list_test.cpp (hunk 5 above)
Comment (mythrocks) [reply cpp/tests/collect_list/collect_list_test.cpp:680]: > A version that takes a vector of values...

Thanks for confirming. I thought that would be useful too. I'll see about adding one to `cudf::test`.

File: cpp/src/groupby/sort/groupby.cu (hunk 4 above)
Comment (mythrocks) [reply cpp/src/groupby/sort/groupby.cu:408]: Hey, @rgsl888prabhu. Good question. 

Null exclusion has not been implemented for `COLLECT()` in general `groupby` aggregations. This PR specifically adds support for  `rolling_window()`.

This assertion is protection against `COLLECT()` being called on non-`rolling_window()` context with `null_policy::EXCLUDE`. The caller might otherwise get the incorrect output.

This assertion will be removed when we add support for `groupby`. Spark will certainly need this.

File: cpp/src/rolling/rolling_detail.cuh (hunk 1 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1070]: > the new ones be an output... more canonical with libcudf style.

I shied from returning `pair<unique_ptr<column>, unique_ptr<column>>`. Let me change this around.

File: cpp/src/rolling/rolling_detail.cuh (hunk 6 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1179]: > Is the gather map here guaranteed to include all elements of the input table?

It is astute of you to notice this. The condition might be rare (especially for calls from Spark's window functions), because each input row's window is likely to touch/overlap with its neighbours'. But it is possible that not all input rows might be selected.

But `rolling_window()` has a generalized version that takes a `preceding_col` and `following_col`. This version of the API allows each row's window to be of arbitrary size, causing rows to be left out, especially at the ends.

Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1179]: P.S. Is `has_nulls()` complex? Should this be avoided if possible?

Comment (vuule) [reply cpp/src/rolling/rolling_detail.cuh:1179]: I meant that the check within `purge_null_entries` could be avoided if all elements are guaranteed to be a part of the aggregation. This is not the case, so I'm resolving the convo :)

File: cpp/tests/collect_list/collect_list_test.cpp (hunk 5 above)
Comment (mythrocks) [reply cpp/tests/collect_list/collect_list_test.cpp]: Sorry for the delay. I've added this utility function.

File: cpp/src/rolling/rolling_detail.cuh (hunk 1 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh]: Fixed, at last. Some of the surrounding code needed moving around to accommodate the change. The null counting needed to be moved out of the function, etc.

**Review by rgsl888prabhu — APPROVED** [review]

File: cpp/include/cudf_test/iterator_utilities.hpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ * iter[7] == true;  // i.e. Valid row at index 7.
+ * iter[8] == false; // i.e. Invalid row at index 8.
+ * iter[9] == false; // i.e. Invalid row at index 9.
+ * @endcode
+ *
+ * @param indices The collection of indices for which the validity iterator
+ *                must return `false` (i.e. null)
+ * @return auto Validity iterator
+ */
+static auto iterator_with_null_at(std::vector<cudf::size_type> const& indices)
```
Comment (jrhemstad) [inline cpp/include/cudf_test/iterator_utilities.hpp:45]: Using a `std::vector` as the input limits this to only that type. Can the input be made an iterator instead?

File: cpp/src/rolling/rolling_detail.cuh [hunk 8]
```diff
@@ -1063,6 +1063,81 @@ struct rolling_window_launcher {
...
+                            input_row_is_null);
+  }
+
+  /**
+   * @brief Purge entries for null inputs from gather_map, and adjust offsets.
+   */
+  std::pair<std::unique_ptr<column>, std::unique_ptr<column>> purge_null_entries(
+    column_view const& input,
+    std::unique_ptr<column> const& gather_map,
+    std::unique_ptr<column> const& offsets,
```
Comment (jrhemstad) [inline cpp/src/rolling/rolling_detail.cuh:1091]: There's almost never a reason to pass a smart pointer by `const&` as a function shouldn't care how the objects lifetime is managed. 
```suggestion
    column_view const& gather_map,
    column_view const& offsets,
```

File: cpp/src/rolling/rolling_detail.cuh [hunk 9]
```diff
@@ -1106,6 +1181,19 @@ struct rolling_window_launcher {
...
+    // If gather_map collects null elements, and null_policy == EXCLUDE,
+    // those elements must be filtered out, and offsets recomputed.
+    auto null_handling = static_cast<collect_list_aggregation*>(agg.get())->_null_handling;
+    if (null_handling == null_policy::EXCLUDE && input.has_nulls()) {
+      auto num_child_nulls = count_child_nulls(input, gather_map, stream);
+      if (num_child_nulls != 0) {
+        auto new_gather_map_and_offsets =
+          purge_null_entries(input, gather_map, offsets, num_child_nulls, stream, mr);
+        gather_map = std::move(new_gather_map_and_offsets.first);
+        offsets    = std::move(new_gather_map_and_offsets.second);
```
Comment (jrhemstad) [inline cpp/src/rolling/rolling_detail.cuh:1193]: Does this not work?
```suggestion
        std::tie(gather_map, offsets) =
          purge_null_entries(input, gather_map, offsets, num_child_nulls, stream, mr);
```

**Review by ajschmidt8 — APPROVED** [review]

File: cpp/include/cudf_test/iterator_utilities.hpp (hunk 7 above)
Comment (vuule) [reply cpp/include/cudf_test/iterator_utilities.hpp:45]: Nice thing about this interface as-is, is that you can simply call it as `iterator_with_null_at({1,3,5})`. With iterators I think you need to use a variable.
May be better to add an overload instead?

Comment (mythrocks) [reply cpp/include/cudf_test/iterator_utilities.hpp:45]: Hmm... Actually, yes. @vuule makes a good point.

Permission to leave this as is?

File: cpp/src/rolling/rolling_detail.cuh (hunk 8 above)
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:1091]: Agreed. This should've been obvious. :/

File: cpp/include/cudf_test/iterator_utilities.hpp (hunk 7 above)
Comment (mythrocks) [reply cpp/include/cudf_test/iterator_utilities.hpp:45]: Actually, as @vuule advises, we should probably have both.

Comment (jrhemstad) [reply cpp/include/cudf_test/iterator_utilities.hpp:45]: Don't use a vector though, use a `host_span`.

File: cpp/include/cudf_test/iterator_utilities.hpp [hunk 10]
```diff
@@ -38,15 +41,43 @@ namespace test {
...
+}
+
+/**
+ * @brief Bool iterator for marking (possibly multiple) null elements in a column_wrapper.
+ *
+ * The returned iterator yields `false` (to mark `null`) at all the specified indices,
+ * and yields `true` (to mark valid rows) for all other indices. E.g.
+ *
+ * @code
+ * auto iter = iterator_with_null_at(std::vector<size_type>{8,9});
```
Comment (jrhemstad) [inline cpp/include/cudf_test/iterator_utilities.hpp:68]: docs need to be updated here and above.

Comment (mythrocks) [reply cpp/include/cudf_test/iterator_utilities.hpp:68]: I'd left the first unchanged since `host_scan`'s constructor is non-explicit, and can accept `std::vector`. But the function template's docs should've been changed.

I'll change both, for clarity.

**Review by vuule — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Comment by mythrocks** [comment]
@gpucibot merge

**Comment by mythrocks** [comment]
Thanks for the reviews, all. I've learnt a bit on this one.
